leaderelection/leaderelection.go

```diff
@@ -0,0 +1,141 @@
+package leaderelection
+
+import (
+	"errors"
+	"os"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+const ConfigMapNameEnv = "CONFIG_LEADERELECTION_NAME"
+
+var errEmptyLeaderElectionConfig = errors.New("empty leader election configuration")
+
+// NewConfigFromMap returns a Config for the given map, or an error.
+func NewConfigFromMap(data map[string]string) (*Config, error) {
+	config := defaultConfig()
+
+	if resourceLock, ok := data["resourceLock"]; ok {
+		config.ResourceLock = resourceLock
+	}
+
+	if leaseDurationStr, ok := data["leaseDuration"]; ok {
+		if leaseDuration, err := time.ParseDuration(leaseDurationStr); err == nil {
+			config.LeaseDuration = leaseDuration
+		}
+	}
+
+	if renewDeadlineStr, ok := data["renewDeadline"]; ok {
+		if renewDeadline, err := time.ParseDuration(renewDeadlineStr); err == nil {
+			config.RenewDeadline = renewDeadline
+		}
+	}
+
+	if retryPeriodStr, ok := data["retryPeriod"]; ok {
+		if retryPeriod, err := time.ParseDuration(retryPeriodStr); err == nil {
+			config.RetryPeriod = retryPeriod
+		}
+	}
+
+	if enabledComponents, ok := data["enabledComponents"]; ok {
+		tokens := strings.Split(enabledComponents, ",")
+		for i, _ := range tokens {
+			str := tokens[i]
+
+			if str == "" {
+				continue
+			}
+
+			config.EnabledComponents[str] = true
+		}
+	}
+
+	return &config, nil
+}
+
+// NewConfigFromConfigMap returns a new Config from the given ConfigMap.
+func NewConfigFromConfigMap(configMap *corev1.ConfigMap) (*Config, error) {
+	if configMap == nil {
+		config := defaultConfig()
+		return &config, nil
+	}
+
+	return NewConfigFromMap(configMap.Data)
+}
+
+// Config represents the leader election config for a set of components
+// contained within a single namespace. Typically these will correspond to a
+// single source repository, viz: serving or eventing.
+type Config struct {
+	ResourceLock  string
+	LeaseDuration time.Duration
+	RenewDeadline time.Duration
+	RetryPeriod   time.Duration
+
+	EnabledComponents map[string]bool
+}
+
+func (c *Config) GetComponentConfig(name string) ComponentConfig {
+	if _, ok := c.EnabledComponents[name]; ok {
+		return ComponentConfig{
+			LeaderElect:   true,
+			ResourceLock:  c.ResourceLock,
+			LeaseDuration: c.LeaseDuration,
+			RenewDeadline: c.RenewDeadline,
+			RetryPeriod:   c.RetryPeriod,
+		}
+	}
+
+	return defaultComponentConfig()
+}
+
+func defaultConfig() Config {
+	return Config{
+		ResourceLock:      "leases",
+		LeaseDuration:     15 * time.Second,
+		RenewDeadline:     10 * time.Second,
+		RetryPeriod:       2 * time.Second,
+		EnabledComponents: map[string]bool{},
+	}
+}
+
+// ComponentConfig represents the leader election config for a single component.
+type ComponentConfig struct {
+	LeaderElect   bool          `json:"leaderElect"`
+	ResourceLock  string        `json:"resourceLock"`
+	LeaseDuration time.Duration `json:"leaseDuration"`
+	RenewDeadline time.Duration `json:"renewDeadline"`
+	RetryPeriod   time.Duration `json:"retryPeriod"`
```

Comment:
I don't think we need the json encodings here?

Reply:
correct, i'll remove them